### ISSUE TITLE
Split query for non-navigation collections

### ISSUE THREAD

**Issue by smitpatel** [dotnet/efcore#21234]
```C#
customer.Select(c => new { c, Orders = c.Orders.Where(o => o.OrderId > 1000) });
customer.Select(c => new { c, OrderDates = c.Orders.Where(o => o.OrderId > 1000).Select(o => o.OrderDate) });
customer.Select(c => new { c, OrderDates = c.Orders.Where(o => o.OrderId > 1000).Select(o => o.OrderDate).Distinct() });
```

**Comment by smitpatel** [comment]
Above queries
1. is easy to do since columns are there, it is somewhat similar to Filtered include.
2. requires adding columns in the projection which are not projection in order to form collections from 2nd query.
3. is not possible in single query mode either since the additional columns required to identify different collections cannot be added due to Distinct.

**Comment by aahmadi458** [comment]
@smitpatel Any updates on this?

**Comment by aahmadi458** [comment]
@smitpatel seems like a think a lot of people need.
23 votes already.
Would you consider planning this for the next release?

**Comment by aahmadi458** [comment]
@smitpatel Know you guys are busy, but would you answer please? Thanks.

**Comment by ajcvickers** [comment]
@aahmadi458 This issue is in the Backlog milestone. This means that it is not planned for the next release (EF Core 5.0). We will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) following the this release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by windhandel** [comment]
@ajcvickers I'm unclear on exactly what has been completed for this feature based on it currently [saying in documentation](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/whatsnew#preview-6) that this has been partly included in the Preview 6.

I was paying particular attention to this feature because of the Logical Reads going exponential which are causing my larger queries to take minutes. This is caused by the cartesian product from the query generated by default from EF Core.

![EFCoreSplitQueryIOExponential](https://user-images.githubusercontent.com/5337357/99460512-d3713b80-28e4-11eb-9ad5-4263bc6bbe26.png)

Could you itemize which of these features are/will be delivered within EF Core v5?

**Comment by ajcvickers** [comment]
@windhandel I believe the patterns not yet supported are the ones listed in this issue. @smitpatel should be able to confirm.

**Comment by smitpatel** [comment]
Using collection navs in Include/ThenInclude can be done with split query. Using them in projection is not yet supported. The issue and documentation captures both.

**Comment by groege** [comment]
I also have the issue having developed with the preview for a while where this worked:

db.MyEntity.Select(x => new MyModel { MyParam = x.MyOtherEntity.Select(y => y.MyString).Distinct()).ToListAsync(ct);

Not a nice surprise finding out it does not actually work in the release version - seems like we have to stick to the prerelease till the end of the year at least until this hopefully is being released with ef6

**Comment by AradAral** [comment]
@smitpatel No updates on this?!

**Comment by smitpatel** [comment]
@AradAral - This issue is planned for EF Core 6.0 the next release of EF Core.

**Comment by AradAral** [comment]
@smitpatel Great to hear. Thanks!

**Comment by Aeroverra** [comment]
I Love the .Net ecosystem but why does it take a whole version change to fix this stuff? Migrated to 3.0 only to find out the Include API was butchered and then upgraded to 5.0 to find out the fix to that still has issues. This is the kind of thing that happens when you fix things that are not broken.

**Comment by johnkwaters** [comment]
I am really upset by this. I have been waiting ever since 2.2 for AsSplitQuery. And now that it's here, it can't handle the use case where it's needed! I fetch a deep object graph off my Trip object with the following includes:

{
"CurrentBreadcrumb",
"Route",
"Route.Driver",
"Route.StartsAt",
"Route.StartsAt.TimeZone",
"Stops",
"Stops.Location",
"Stops.Location.TimeZone",
"Stops.Location.Address",
"Stops.Location.Address.State",
"Driver",
"Legs",
"Stops.Activities",
"Stops.Activities.ActivityType",
"Stops.Activities.Properties",
"Stops.Activities.SurveyResult",
"Stops.Activities.Survey",
"TripErrors",
"TripErrors.Stop",
"Metrics"
});

2.2 did it fast. 3.0. 3.1 and 5.0 use tons of memory and take a minute. SplitQuery wont process it. Am I supposed to hand code all of the parts of assembling this tree? What's the point of an ORM then? Why did you remove this feature and not add back any compatibility?

**Comment by imt-s1897** [comment]
I find it strange that EF Core 2.2 is better than the newer versions. Insane. At the moment I have to use EF Plus whenever EF Core 5 just sucks. This is one of these situations.

**Comment by goldsam** [comment]
Kind of a bid deal that OData is broken without this

**Comment by jsheetzati** [comment]
We purposely skipped EFCore 3.x because of the single vs split query difference between 2.x and 3.x. Now, I am trying to migrate to 5.x and enabled SplitQuery globally. However, because of this open issue, we are finding ourselves getting hit with random runtime exceptions. 

Does `Microsoft.EntityFrameworkCore.Analyzers` detect this situation at compile time? We are looking for a way to detect this situation rather than waiting for runtime exceptions.

**Comment by AradAral** [comment]
Sorry, is this implemented? Because I'm on EF Core 6.0.0-preview.3.21201.2 and I'm still getting the error:

![image](https://user-images.githubusercontent.com/26527405/116597230-981a7480-a93a-11eb-8a2b-13016e11d958.png)
